apps/examples/smartfs/buffer_optimization/buffer_optimization_main.c

```diff
@@ -0,0 +1,449 @@
+	if (ioctl(frt_fd, TCIOC_GETSTATUS, (unsigned long)(uintptr_t)&time_2) < 0) {
+		fprintf(stderr, "ERROR: Failed to get Free Run Timer status: %d\n", errno);
+		return -1;
+	}
+
+	time_read = time_2.timeleft - time_1.timeleft;
+	print_test_result("reading", size, READ_TEST_ITR, 1024, time_read);
+
+	ret = stat("/mnt", &st);
+	if (ret < 0) {
+		printf("Stat to get optimal read buffer size failed, Errno = %d\n", errno);
+		return -1;
+	}
+	buf_size = st.st_blksize;
+
+	if (ioctl(frt_fd, TCIOC_GETSTATUS, (unsigned long)(uintptr_t)&time_1) < 0) {
+		fprintf(stderr, "ERROR: Failed to get Free Run Timer status: %d\n", errno);
+		return -1;
+	}
+
+	ret = read_test(buf_size);
+	if (ret != OK) {
+		printf("Reading with buffer size %d failed\n", buf_size);
+		return ret;
+	}
+
+	if (ioctl(frt_fd, TCIOC_GETSTATUS, (unsigned long)(uintptr_t)&time_2) < 0) {
+		fprintf(stderr, "ERROR: Failed to get Free Run Timer status: %d\n", errno);
+		return -1;
+	}
+
+	time_read = time_2.timeleft - time_1.timeleft;
+	printf("TEST WITH OPTIMIZED BUFFER\n");
+	print_test_result("reading", size, READ_TEST_ITR, buf_size, time_read);
+
+	printf("================================== READ TESTS COMPLETE ==================================\n\n");
+
+	return OK;
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+#ifdef CONFIG_BUILD_KERNEL
+int main(int argc, char *argv[])
+#else
+int buffer_optimization_main(int argc, char *argv[])
+#endif
+{
+	int ret;
+	int size = 8;
+	int frt_fd = -1;
+	int timer_devno = 0;
+	char timer_path[_POSIX_PATH_MAX];
+
+	if (argc > 1) {
+		timer_devno = atoi(argv[1]);
+		if (argc == 3) {
+			size = atoi(argv[2]);
+		}
+	} else {
+		printf("\nCorrect test command usage is: smartfs_buf_opt <Timer Device No.> <File Size in Kb>\n");
+		printf("-> Read test example will read the file back to back, one buffer size at a time\n");
+		printf("-> Append test example will append 10Kb data to the file, one buffer size at a time\n");
+		return -1;
+	}
+
+	snprintf(timer_path, _POSIX_PATH_MAX, TIMER_DEVNAME, timer_devno);
+	printf("Timer path = %s\n", timer_path);
+
+	ret = create_test_file(size * 1024);
+	if (ret != OK) {
+		printf("Unable to complete test\n");
+		return -1;
+	}
+
+	frt_fd = open(timer_path, O_RDONLY);
+	if (frt_fd < 0) {
+		fprintf(stderr, "ERROR: Failed to open Free Run Timer: %d\n", errno);
+		return -1;
+	}
+
+	if (ioctl(frt_fd, TCIOC_SETMODE, MODE_FREERUN) < 0) {
+		fprintf(stderr, "ERROR: Failed to set Free Run Timer: %d\n", errno);
+		return -1;
+	}
+	if (ioctl(frt_fd, TCIOC_START, TRUE) < 0) {
+		fprintf(stderr, "ERROR: Failed to start Free Run Timer: %d\n", errno);
+		return -1;
+	}
+
+	ret = read_test_util(frt_fd, size);
+	if (ret != OK) {
+		printf("Read test failed!!!\n");
+		return -1;
+	}
+
+	ret = append_test_util(frt_fd, size);
+	if (ret != OK) {
+		printf("Append test failed\n");
+		return -1;
+	}
+
```

Comment:
Why don't you add create_test_until()?

Reply:
File creation will have no effect on the size of buffer. The file creation we used was as  :
 ```
 Let's say creating 32kb file.
 ret  = write (fd, buffer, 32768);
 ```
So, this has no role of optimized buffer.

Reply:
Regarding the comment that I left in the function, create_test_file(), we can use an optimized buffer size as write buffer sizes. I'm still wondering if there is no gains even if we use st.st_blksize instead of 1024.

Reply:
@dineshdas-samsung Your example doesn't have any gains because the buffer size is not optimized, but just big enough to accommodate the entire file.

Reply:
I strongly agree with your point @davidfather . And also we have gains with optimized buffer but If we create a file with optimized buffer size (say 1013) for any file size (>1013) it will be case of append. And for that we already have the append_test_util().

Reply:
We expect gains if we compare performance of buffer size = 1024 vs buffer size = st_blksize.
However, we also feel that this would change the nature of the test from direct file creation to buffer-by-buffer append-create.